# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
path="/home/tdraths/SB Assignments/Big Mountain Guided Capstone/data"
os.chdir(path)
df = pd.read_csv('step3_output.csv')

df.head()

Unnamed: 0                 Name    state  summit_elev  vertical_drop  \
0           0       Alyeska Resort   Alaska         3939           2500   
1           1  Eaglecrest Ski Area   Alaska         2600           1540   
2           2     Hilltop Ski Area   Alaska         2090            294   
3           3     Arizona Snowbowl  Arizona        11500           2300   
4           4  Sunrise Park Resort  Arizona        11100           1800   

   base_elev  trams  fastEight  fastSixes  fastQuads  ...  SkiableTerrain_ac  \
0        250      1        0.0          0          2  ...             1610.0   
1       1200      0        0.0          0          0  ...              640.0   
2       1796      0        0.0          0          0  ...               30.0   
3       9200      0        0.0          1          0  ...              777.0   
4       9200      0        0.0          0          1  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         2  
4          78.0              104.0            80.0         2  

[5 rows x 28 columns]

In [3]:
df.drop(['Unnamed: 0'], axis=1, inplace=True)

In [4]:
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         2  
4          78.0              104.0            80.0         2  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

#NOTE FROM TOM: I could not get the code in Aiden's article to work (after making the changes needed for my dataset). Instead, I looked up a way that is longer, but allowed me to work step-by-step through the creation of the variable, the concat with df, and the drop of original 'state'

In [5]:
#Create dummy dataframe
dummy = pd.get_dummies(df['state'])
dummy.head()

Alaska  Arizona  California  Colorado  Connecticut  Idaho  Illinois  \
0       1        0           0         0            0      0         0   
1       1        0           0         0            0      0         0   
2       1        0           0         0            0      0         0   
3       0        1           0         0            0      0         0   
4       0        1           0         0            0      0         0   

   Indiana  Iowa  Maine  ...  Rhode Island  South Dakota  Tennessee  Utah  \
0        0     0      0  ...             0             0          0     0   
1        0     0      0  ...             0             0          0     0   
2        0     0      0  ...             0             0          0     0   
3        0     0      0  ...             0             0          0     0   
4        0     0      0  ...             0             0          0     0   

   Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0        0         0           0              0          0        0  
1        0         0           0              0          0        0  
2        0         0           0              0          0        0  
3        0         0           0              0          0        0  
4        0         0           0              0          0        0  

[5 rows x 35 columns]

In [6]:
#Add dummy into the original df
df = pd.concat([df, dummy], axis = 1)

In [7]:
#Drop original 'state' column
df.drop('state', axis=1, inplace=True)

In [8]:
df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          0          2     2       0  ...             0   
1        0.0          0          0     0       0  ...             0   
2        0.0          0          0     0       1  ...             0   
3        0.0          1          0     2       2  ...             0   
4        0.0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [9]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing as p

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = p.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [10]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [11]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [12]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)
y_pred

array([ 57.286938  ,  91.24201612,  78.26935987,  76.43732862,
        70.15607862,  76.2010005 ,  44.98615675,  46.91389112,
        93.03498487,  83.15803175,  51.34553175,  79.5135005 ,
        50.50373487,  71.76545362,  62.927563  ,  92.63654737,
        37.0291255 ,  72.74982862,  61.07795362,  41.30842237,
       113.693188  ,  36.59357862,  53.20685987,  48.94514112,
        48.56428175,  62.474438  ,  75.66584425,  81.31623487,
        64.61310987,  66.2478755 ,  35.15888624,  60.79670362,
        78.8572505 ,  56.72199659,  75.17170362,  49.06428175,
        23.24592237,  29.03498487,  62.2322505 ,  37.93732862,
        48.34748487,  60.724438  ,  85.3416255 ,  71.74592237,
        71.29279737,  57.00959425,  64.21076612, 165.84748487,
        38.07856397,  62.958813  ,  68.91804151,  58.63654737,
        52.73420362,  46.11701612,  59.12873487,  42.79523878,
        87.40314893,  38.6853755 , 119.04084425, 119.24201612,
        56.3260005 ,  56.11701612,  41.94123487,  79.80

## Review Model Outcomes — Iterate over additional models as needed

In [13]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [14]:
explained_variance_score(y_test, y_pred)

0.9289936877696242

In [15]:
mean_absolute_error(y_test, y_pred)

5.192127035848631

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [16]:
lm.intercept_

64.12092237065369

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [17]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeffvalues = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

coeffvalues.sort_values(by=['Coefficient'], inplace=True, ascending=False)
coeffvalues

Coefficient
total_chairs       2.848842e+13
fastQuads          1.080003e+13
surface            1.011881e+13
double             8.917074e+12
triple             7.954642e+12
quad               6.446943e+12
New York           6.282746e+12
Michigan           5.929206e+12
Colorado           5.223974e+12
California         5.112145e+12
Pennsylvania       4.878331e+12
Wisconsin          4.498204e+12
New Hampshire      4.498204e+12
Vermont            4.362297e+12
Minnesota          4.221063e+12
Utah               4.073949e+12
Montana            3.920293e+12
Idaho              3.920293e+12
Massachusetts      3.759291e+12
Oregon             3.589957e+12
Washington         3.589957e+12
New Mexico         3.411050e+12
Maine              3.411050e+12
Wyoming            3.220974e+12
fastSixes          3.201673e+12
North Carolina     2.798095e+12
trams              2.750966e+12
Ohio               2.558238e+12
Connecticut        2.558238e+12
Nevada             2.291675e+12
Illinois           2.291675e+12
Virginia           2.291675e+12
West Virginia      2.291675e+12
Iowa               1.987691e+12
Alaska             1.987691e+12
Arizona            1.625422e+12
South Dakota       1.625422e+12
Indiana            1.625422e+12
Missouri           1.625422e+12
New Jersey         1.625422e+12
Rhode Island       1.151098e+12
Maryland           1.151098e+12
Tennessee          1.151098e+12
fastEight          2.704469e+11
summit_elev        2.133062e+01
AdultWeekday       1.926074e+01
base_elev          1.863165e+01
vertical_drop      6.128569e+00
clusters           3.177246e+00
Snow Making_ac     8.717041e-01
Runs               8.684082e-01
NightSkiing_ac     7.265625e-01
LongestRun_mi      4.481201e-01
yearsOpen          3.880615e-01
averageSnowfall    3.813477e-01
daysOpenLastYear   3.000488e-01
TerrainParks       2.514648e-01
projectedDaysOpen  1.870117e-01
SkiableTerrain_ac  1.293945e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [18]:
path="/home/tdraths/SB Assignments/Big Mountain Guided Capstone/data"
os.chdir(path)
df = pd.read_csv('step3_output.csv')
df.drop('Unnamed: 0', axis=1, inplace=True)
df.drop('state', axis=1, inplace=True)

In [19]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing as p

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = p.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [20]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [21]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [22]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)
y_pred

array([ 57.54264813,  84.64405735,  81.90254345,  78.25143251,
        67.02169376,  80.52122063,  37.67561383,  47.44177486,
        88.60234084,  85.69760044,  51.86112974,  79.37317352,
        53.73307888,  65.4934766 ,  63.25388774,  88.99415683,
        37.51086466,  76.86206382,  64.54663118,  43.33454471,
       108.23975892,  37.08877369,  53.50995344,  49.89380793,
        50.56050876,  63.46056286,  66.14642527,  75.70165973,
        64.09669442,  64.49467462,  39.44612168,  52.31202126,
        77.85788784,  56.96632852,  75.13772447,  54.13458698,
        25.26168477,  27.88897267,  55.23487289,  38.97139193,
        49.88294317,  64.43971969,  88.47311433,  66.72210688,
        68.50384229,  53.46591318,  62.63300763, 168.21231673,
        30.45811752,  63.59214142,  63.06192392,  60.03125673,
        53.55060837,  48.51708847,  61.26114406,  42.80345384,
        87.10851926,  31.17596827, 121.67363068, 115.19122369,
        57.65978466,  54.39141304,  39.41874764,  86.66

In [23]:
explained_variance_score(y_test, y_pred)

0.9217534455931519

In [24]:
mean_absolute_error(y_test, y_pred)

5.559550694118946

In [25]:
lm.intercept_

64.05723991598244

In [26]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeffvalues = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

coeffvalues.sort_values(by=['Coefficient'], inplace=True, ascending=False)
coeffvalues

Coefficient
AdultWeekday         20.160108
summit_elev          13.246478
base_elev             9.986536
vertical_drop         4.472549
averageSnowfall       1.843946
quad                  1.515646
triple                1.366897
surface               1.224561
daysOpenLastYear      1.113466
Runs                  1.097200
fastQuads             0.659281
total_chairs          0.650527
Snow Making_ac        0.616768
projectedDaysOpen     0.436641
clusters              0.377650
NightSkiing_ac        0.332057
double                0.324386
fastSixes             0.313111
LongestRun_mi         0.301440
yearsOpen             0.282784
trams                 0.261066
SkiableTerrain_ac     0.246429
TerrainParks          0.244644
fastEight             0.066525

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [27]:
df.drop('summit_elev', axis=1, inplace=True)
df.drop('base_elev', axis=1, inplace=True)

In [28]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing as p

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = p.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [29]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [30]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [31]:
y_pred = model.predict(X_test)
y_pred

array([ 58.49097745,  83.69644043,  81.88066721,  78.50846501,
        67.16715672,  80.15223856,  37.6013702 ,  47.38355601,
        88.66050622,  85.99558745,  51.78208029,  79.54892498,
        54.17100363,  65.32072115,  63.29263001,  89.28532728,
        40.9059166 ,  76.51651854,  65.74293105,  43.181091  ,
       107.71706725,  37.06474607,  53.21635294,  49.89831232,
        50.36816834,  64.18737561,  66.66361638,  72.34913861,
        64.53372359,  64.6135482 ,  39.87303954,  51.55677623,
        78.55922124,  56.7912267 ,  75.58611254,  54.14060696,
        25.0604635 ,  28.18464558,  54.91020944,  38.74501655,
        49.98409042,  64.81453219,  88.89657329,  67.04424543,
        69.17267612,  53.28761524,  62.16872653, 168.15752793,
        30.48430439,  63.12673696,  62.67212689,  60.82218431,
        53.41310003,  48.02865227,  61.25117716,  43.30285297,
        85.7915807 ,  31.79268537, 121.94955344, 114.83131137,
        57.48756905,  53.57575719,  39.32367482,  86.57

In [32]:
explained_variance_score(y_test, y_pred)

0.9279292510972763

In [33]:
mean_absolute_error(y_test, y_pred)

5.430951704536332

In [34]:
lm.intercept_

64.06760502891679

In [35]:
coeffvalues = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

coeffvalues.sort_values(by=['Coefficient'], inplace=True, ascending=False)
coeffvalues

Coefficient
AdultWeekday         20.113666
averageSnowfall       1.862646
quad                  1.496367
vertical_drop         1.358216
triple                1.331662
Runs                  1.280110
surface               1.208159
daysOpenLastYear      1.056232
fastQuads             0.767236
clusters              0.736239
total_chairs          0.686903
Snow Making_ac        0.510810
NightSkiing_ac        0.430684
fastSixes             0.374998
LongestRun_mi         0.330780
projectedDaysOpen     0.321640
double                0.291174
yearsOpen             0.283403
SkiableTerrain_ac     0.278504
trams                 0.202320
TerrainParks          0.176870
fastEight             0.101313

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9289936877696242 | 5.192127035848631 |-|
| Model 2. | 0.9217534455931519 | 5.559550694118946 |'state'|
| Model 3. | 0.9279292510972763 | 5.430951704536332 |'state','summit_elev','base_elev'|

Model Selection:

I would choose Model 3 as the best model to explain the features driving ski resort prices. Because we dropped the state and two elevation columns from the dataframe before running the model, Model 3 does not consider features that the resort cannot control: where it is located, and the mountain's size. In the top 10 features in Model 3, we see some things that would be more difficult for management to change, like the number of lifts and average snowfall, but we also see some big ones that can be changed, notably the weekday price, runs, and days open last year.  While you cannot go back in time and re-open the resort last year, you can assure that you have a few more days open this year by using snowmaking where available to extend the season.